File: 백준/Gold/1167. 트리의 지름/트리의 지름.py
import sys
data_temp = sys.stdin if sys.platform == 'linux' else open('입력.txt', 'r')
input_data = data_temp.read().splitlines()
from collections import deque

def solution (data) :
  n = int(data.pop(0))
  
  graph = [[] for _ in range(n+1)]
  
  for _str in data :
    [node, *lis] = list(map(int, _str.split()))[:-1]
    
    for i in range(0, len(lis), 2) :
      graph[node].append([lis[i+1], lis[i]])

  def BFS (node) : 
    visited = [-1 for _ in range(n+1)]
    visited[node] = 0
    queue = deque([])
    queue.append([0, node])
    _max = [0,0]
    
    while queue :
      dist, node = queue.popleft()
      for nextDist, nextNode in graph[node] :
        if visited[nextNode] == -1 :
          totalDist = nextDist + dist
          queue.append([totalDist, nextNode])
          visited[nextNode] = totalDist
          if totalDist > _max[1] :
            _max[0] = nextNode
            _max[1] = totalDist
            
    return _max
    
  
  leaf = BFS(1)
  result = BFS(leaf[0])[1]
  print(result)
    
solution(input_data)

File: 백준/Gold/1967. 트리의 지름/트리의 지름.js
const fs = require("fs");
const input = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "입력.txt")
  .toString()
  .trim()
  .split("\n");

function solution(data) {
  const n = +data.shift();
  const arr = data.map((el) => el.split(" ").map(Number));

  const graph = Array.from({ length: n + 1 }, () => []);
  arr.forEach((el) => {
    const [s, e, w] = el;
    graph[s].push([w, e]);
    graph[e].push([w, s]);
  });

  function BFS(node) {
    const visited = Array.from({ length: n + 1 }, () => false);
    const distances = Array.from({ length: n + 1 }, () => 0);
    const queue = [[node, 0]];

    while (queue.length) {
      const [node, dist] = queue.shift();

      if (visited[node]) continue;
      visited[node] = true;
      distances[node] = dist;

      graph[node].forEach(([nextDist, nextNode]) => {
        const totalDist = dist + nextDist;
        queue.push([nextNode, totalDist]);
      });
    }
    const maxValue = Math.max(...distances);
    const maxIdx = distances.indexOf(maxValue);
    return [maxIdx, maxValue];
  }

  const leaf = BFS(1);
  const result = BFS(leaf[0])[1];
  console.log(result);
}

solution(input);
